grove/arbitrary_state/arbitrary_state.py

```diff
@@ -0,0 +1,214 @@
+        rotation_cnots[-1] = i
+        rotation_cnots = rotation_cnots + rotation_cnots
+    return rotation_cnots
+
+
+def create_arbitrary_state(vector, qubits=None):
+    """
+    This function makes a program that can generate an arbitrary state.
+
+    Applies the methods described in references above.
+
+    Given a complex vector :math:`\\mathbf{a}` with components :math:`a_i`
+    (:math:`i` ranging from :math:`0` to :math:`N-1`),
+    produce a program that takes in the state :math:`\\vert 0 \\rangle`
+    and outputs the state
+
+    .. math::
+
+        \\sum_{i=0}^{N-1}\\frac{a_i}{\\vert \\mathbf{a}\\vert} \\vert i\\rangle
+
+    where :math:`i` is given in its binary expansion.
+
+    :param 1darray vector: the vector to put into qubit form.
+    :param list(int) qubits: Which qubits to encode the vector into.
+                             Must contain at least the minimum
+                             number of qubits :math:`n` needed for all elements
+                             of vector to be present as a coefficient in the
+                             final state. If more than :math:`n` are provided,
+                             only the first :math:`n` will be used.
+                             If no list is provided, the default will be
+                             qubits :math:`0, 1, \ldots, n-1`.
+    :return: a program that takes in :math:`\\vert 0\\rangle^{\otimes n}`
+             and produces a state that represents this vector, as
+             described above.
+    :rtype: Program
+    """
+    vec_norm = vector / np.linalg.norm(vector)
+    n = max(1, int(np.ceil(np.log2(len(vec_norm)))))  # number of qubits needed
+
+    if qubits is None:
+        qubits = range(n)
+
+    N = 2 ** n  # number of coefficients
+    while len(vec_norm) < N:
+        vec_norm = np.append(vec_norm, 0)  # pad with zeros
+
+    magnitudes = map(np.abs, vec_norm)
+    phases = map(np.angle, vec_norm)
+
+    # because conceptually this algorithm starts with the end state
+    # and makes it into the |0> state,
+    # the gates will be constructed in reverse order
+    reversed_gates = []
+
+    # matrix that converts angles of uniformly controlled rotation
+    # to angles of uncontrolled rotations
+    # at first, all qubits except for the 0 qubit act as controls
+    M = get_uniformly_controlled_rotation_matrix(n - 1)
+
+    # generate the positions of the controls for the CNOTs
+    # at first, all qubits except for the 0 qubit act as controls
+    rotation_cnots = get_cnot_control_positions(n - 1)
+
+    for step in xrange(n):
+        # will hold the angles for controlled rotations
+        # in the phase unification and probability unification steps,
+        # respectively
+        z_thetas = []
+        y_thetas = []
+
+        # will hold updated phases and magnitudes after rotations
+        new_phases = []
+        new_magnitudes = []
+
+        for i in xrange(0, len(phases), 2):
```

Comment:
This section can be more modular to make for easier reading and testing.  You could have an extra function that takes a set of phases and returns two lists `new_phases` and `new_magnitudes`.